### PR TITLE
EDLY_1280 Filter discovery courses/course runs API’s WRT organization

### PR DESCRIPTION
**Pull request by iam-mhaseeb**
**Description:** This PR shows how we can filter courses, the course runs using course discovery API with respect to the organization. 

**JIRA:** 
https://edlyio.atlassian.net/browse/EDLY-1280

**Testing instructions:**
Tested filteration with following endpoints to filter based on organization:
- GET /api/v1/course_runs/?organization=edly
- GET /api/v1/course_runs/contains/?organization=edly
- GET /api/v1/course_runs/{key}/?organization=edly
- GET /api/v1/courses/?organization=edly
- GET /api/v1/courses/{key}/?organization=edly
- GET /api/v1/search/all/?org=edly
- GET /api/v1/search/all/details/?org=edly
- GET /api/v1/search/all/facets/?org=edly
- GET /api/v1/search/all/{key}/?org=edly
- GET /api/v1/search/course_runs/?org=edly
- GET /api/v1/search/course_runs/details/?org=edly
- GET /api/v1/search/course_runs/facets/org=edly
- GET /api/v1/search/courses/?org=edly
- GET /api/v1/search/courses/details/?org=edly
- GET /api/v1/search/courses/facets/?org=edly
- GET /api/v1/search/courses/{key}/?org=edly

**Merge checklist:**
- I think this pull request won't merge yet.

**Post merge:**
- [ ] Delete working branch (if not needed anymore)

**Note:** Following screenshot shows filtration in postman:
Screenshot 2020-05-04 at 12 32 38 PM

### CODE REVIEW

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 1]
```diff
@@ -124,6 +125,12 @@ def list(self, request, *args, **kwargs):
               type: integer
               paramType: query
               multiple: false
+            - name: organization
+              description: Filter course runs by sub organizations or edx organization.
```
Comment (zubair-arbi) [inline course_discovery/apps/api/v1/views/course_runs.py]: Filter course runs by edly sub organization or edx organization.

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 2]
```diff
@@ -124,6 +125,12 @@ def list(self, request, *args, **kwargs):
...
+              required: false
```
Comment (zubair-arbi) [inline course_discovery/apps/api/v1/views/course_runs.py:130]: I believe for our case the organization filter will be required. So that we can make sure that every service is using the courses data related.

**Review by zubair-arbi — CHANGES_REQUESTED** [review]
@iam-mhaseeb as we have decided that we set a new cookie `edly-user-info` (https://edlyio.atlassian.net/browse/EDLY-1292) at user login and it will be set at a common domain (using wildcard name) so that its shard across all edly cloud services. Also the encode/decode secret key will be shared across all edly cloud services. 

So I think we will not need to append organization query parameter in  the API URL's in edly services, e.g. WordPress, ECOM etc instead we can just decode the  cookie `edly-user-info` to get information like `edly-org` (slug) and `edly-sub-org` (edx organization short_name) on all API calls automatically. 

We will need add a new edly-utils file where we will have a method `get_edly_sub_organization(request)` and this method will fetch the `edly-sub-org` value from the cookie `edly-user-info` (with a cache of ~30 minutes) and this method can be called from DRF serializers or views.

@ziafazal Please provide your feedback on this approach.

**Review by ziafazal — APPROVED** [review]
👍

File: course_discovery/apps/api/v1/views/course_runs.py [hunk 3]
```diff
@@ -13,6 +13,8 @@
 from course_discovery.apps.course_metadata.constants import COURSE_RUN_ID_REGEX
 from course_discovery.apps.course_metadata.models import CourseRun
 
+from course_discovery.apps.api.edly_utils import get_edly_sub_organization
```
Comment (ziafazal) [inline course_discovery/apps/api/v1/views/course_runs.py]: @iam-mhaseeb where I can find the implementation of `get_edly_sub_organization`?

Comment (iam-mhaseeb) [reply course_discovery/apps/api/v1/views/course_runs.py]: @ziafazal It was missed in the previous commit. For now, it's dummy function it's actual implementation would be done after cookie setting flow when the user logs in.

**Comment by iam-mhaseeb** [comment]
@ziafazal @zubair-arbi  Just as heads up, as we are now moving away from passing the query parameter in API URL so changes in these PR are almost stale. For now, I'm keeping this PR open to constantly keep getting feedback on new changes, later I"ll create fresh PR.

**Comment by zubair-arbi** [comment]
@iam-mhaseeb as discussed we might need to use both approaches (using URL base filtering and using cookie base filtering)

Getting￼edly-sub-or￼ value from the cookie￼edly-user-inf￼ is a clean solution but this will not work for anonymous users where the user is not logged in and the cookie￼edly-user-inf￼ isn’t set

For WordPress where we query through Ajax we will be using URL based filtering (for anonymous users) and on ECOM it will be better if we filter courses/course run through edly-sub-org value from the cookie edly-user-info.

In short for now we will be using both approaches.

if the URL contains organization filter then we will filter courses/course run (listing and search) based on that query parameter

if URL contains no query parameter with the name organization then we will try to get edly-sub-org value from the cookie edly-user-info (for logged in users case only)

**Comment by iam-mhaseeb** [comment]
Closing this pull request as this was for just r&d.
